go-controller/pkg/node/gateway.go

```diff
@@ -0,0 +1,187 @@
+package node
+
+import (
+	"fmt"
+	"net"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/informer"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/kube"
+	util "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+	kapi "k8s.io/api/core/v1"
+	"k8s.io/klog"
+)
+
+// Gateway responds to Service and Endpoint K8s events
+// and programs OVN gateway functionality.
+// It may also spawn threads to ensure the flow tables
+// are kept in sync
+type Gateway interface {
+	informer.ServiceAndEndpointsEventHandler
+	Init() error
+	Run(<-chan struct{})
+}
+
+type gateway struct {
+	loadBalancerHealthChecker informer.ServiceAndEndpointsEventHandler
+	portClaimWatcher          informer.ServiceEventHandler
+	nodePortWatcher           informer.ServiceEventHandler
+	localPortWatcher          informer.ServiceEventHandler
+	openflowManager           *openflowManager
+	initFunc                  func() error
+}
+
+func (g *gateway) AddService(svc *kapi.Service) {
```

Comment:
I'm not sure I get the utility here. Informers already mux out to multiple subscribers; what's wrong with having the individual components (e.g. localport code) just set up their handlers directly?

Reply:
The motivation for this refactor is to insert a [workqueue ](https://godoc.org/k8s.io/client-go/util/workqueue) between the informers and the event subscribers. I'm hopeful that it will remove the need for us to maintain `pkg/factory` and will provide us with a nice performance gain, alongside some additional benefits (i.e prometheus metrics for queues, ability to re-renqueue work etc...)

The trade-off is that we need to handle event dispatch to each individual component - i.e. we've pulled something off the queue... now we need it to go to X, Y and Z in either series or parallel.

What's quite nice about making this explicit is that it makes it easier to understand what's going on.
Another side-effect is that the component code becomes easier to test - we can test it's AddService, DeleteService functiions in isolation